### PR TITLE
Insert `ON CONFLICT`

### LINKED ISSUE #88: `UPSERT` statement support
In [sqlite] and [postgres], it is possible to use the `ON CONFLICT` clause for `INSERT` statements in order to do an [upsert]

### Example

As an example suppose you have the following table: 
```sql
CREATE TABLE customers (
	customer_id serial PRIMARY KEY,
	name VARCHAR UNIQUE,
	email VARCHAR NOT NULL,
	active bool NOT NULL DEFAULT TRUE
);
```

and you wish to insert without having errors and it is safe to update if it already exists in the table, this can all be handled in sql in the following way:
```sql
INSERT INTO customers (name, email)
VALUES('Microsoft','hotline@microsoft.com') 
ON CONFLICT (name) 
DO 
   UPDATE SET email = EXCLUDED.email;
```

_Note: all examples are taken directly from the [upsert] link (postgresqltutorial.com)._

[sqlite]: https://sqlite.org/lang_conflict.html
[postgres]: https://www.postgresql.org/docs/13/sql-insert.html#SQL-ON-CONFLICT
[upsert]: https://www.postgresqltutorial.com/postgresql-upsert/

### Api suggestion

This is meant as a suggestion of what could be nice to  have as an api for this, this is by no means what might end up being implemented. It might not be possible to implement as written below.

The clause in both cases has two parameters, a `target` column on which a constraint is generally applied, and a `action`.

```rust
use sea_query::{Query, ConflictAction};

// upsert example
Query::insert()
    .into_table(Customers::Table)
    .columns([Customers::Name, Customers::Email])
    .on_conflict(Customers::Name, ConflictAction::Update(Query::update().set(Customers::Email))

// ignore example
Query::insert()
    .into_table(Customers::Table)
    .columns([Customers::Name, Customers::Email])
    .on_conflict(Customers::Name, ConflictAction::Nothing)
```

### Alternatives

An alternative to supporting the `ON CONFLICT` clause would be to implement the `UPSERT` statement directly which is supported in 
* Postgresql version >= 9.5
* Sqlite version >= 3.24.0 (2018-06-04)
* mysql version >= none

But seeing as `ON CONFLICT` is more powerful and seems to be the best practice in Postgresql anyways. I'm not sure this would be a viable alternative.

### Limitations

Seeing as `ON CONFLICT` is not standard SQL this can lead to issues, namely the fact that [MySql handles upserts differently][mysql]. Where the `ON CONFLICT(target) DO UPDATE` would actually be `ON DUPLICATE KEY UPDATE` and it doesnt seem like you can add a target. The previous was the easy one, the others such as `DO NOTHING` changes the statement entirely to be `INSERT IGNORE ...` I'm not sure how this can be implemented cleanly.

[mysql]: https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html

### PR DESCRIPTION
**Pull request by billy1624**
## PR Info

- Closes #88

## Adds

- Support `ON CONFLICT`

### CODE REVIEW

File: src/query/on_conflict.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+/// Represents ON CONFLICT (upsert) actions
+#[derive(Debug, Clone)]
+pub enum OnConflictAction {
+    /// Update the column values of existing row instead of insert
+    UpdateColumns(Vec<DynIden>),
+}
+
+impl OnConflict {
+    pub fn new() -> Self {
```
[the remark this answers is not in the document]
Comment (billy1624) [reply src/query/on_conflict.rs:24]: Yes, you're correct

[the remark this answers is not in the document]
Comment (billy1624) [reply src/query/on_conflict.rs:24]: For MySQL, 

https://github.com/SeaQL/sea-query/blob/5440ba1208b12fd7db95ceaa9677768cf3b8c3fb/tests/mysql/query.rs#L1089-L1111

[the remark this answers is not in the document]
Comment (tyt2y3) [reply src/query/on_conflict.rs:24]: @Progdrasil makes sense

**Comment by billy1624** [comment]
I think we still need to add two more `OnConflictAction`
- Update by `Value`
- Update by `SimpleExpr`

**Comment by billy1624** [comment]
> I think we still need to add two more `OnConflictAction`
> 
> * Update by `Value`
> * Update by `SimpleExpr`

Added on ae692a98dca475ae7290eec0d285bc29ded54404. Ready for review

File: src/backend/query_builder.rs [hunk 2]
```diff
@@ -1079,6 +1081,116 @@ pub trait QueryBuilder: QuotedBuilder {
...
+    fn prepare_on_conflict_action(
+        &self,
+        on_conflict_action: &Option<OnConflictAction>,
+        sql: &mut SqlWriter,
+        collector: &mut dyn FnMut(Value),
+    ) {
+        if let Some(action) = on_conflict_action {
+            match action {
+                OnConflictAction::DoNothing => {
+                    write!(sql, " DO NOTHING").unwrap();
```
Comment (tyt2y3) [inline src/backend/query_builder.rs:1135]: I think having the prefix space embedded here is bad

**Pull request merged**